File: IClient.vb
Imports System.Collections.Generic
Imports System.IO
Imports System.Threading.Tasks
Imports ISisYandexSDK.JSONyandex
Imports ISisYandexSDK.YClient
Imports ISisYandexSDK.YdXutilities


Public Interface IClient



#Region "file"
    Function ListAllFiles(Optional MediaType As List(Of String) = Nothing, Optional Fields As Fields = Nothing, Optional Limit As Integer = 20, Optional Offset As String = Nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As String = Nothing, Optional Sort As Sort = Nothing) As Task(Of JSONAllFiles)
    Function CopyFile(SourceFile As String, DestinationFolder As String, FileName As String, Optional Fields As Fields = Nothing, Optional Overwriting As Boolean = False) As Task(Of JSONLink)
    Function MoveFile(SourceFile As String, DestinationFolder As String, FileName As String, Optional Fields As Fields = Nothing, Optional Overwriting As Boolean = False) As Task(Of JSONLink)
    Function DeleteFile(FilePath As String, Optional Fields As Fields = Nothing) As Task(Of JSONLink)
    Function TrashFile(FilePath As String, Optional Fields As Fields = Nothing) As Task(Of JSONLink)
    Function DownloadFile(FullFilePath As String, FileSaveDir As String, FileName As String, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional _proxi As ProxyConfig = Nothing, Optional TimeOut As Integer = 60, Optional token As Threading.CancellationToken = Nothing) As Task
    Function DownloadFileAsStream(FullFilePath As String, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional _proxi As ProxyConfig = Nothing, Optional TimeOut As Integer = 60, Optional token As Threading.CancellationToken = Nothing) As Task(Of IO.Stream)
    Function UploadLocalFile(FileToUpload As Object, FileName As String, DestinationFolder As String, UploadType As UploadTypes, Optional Overwriting As Boolean = False, Optional Fields As Fields = Nothing, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional _proxi As ProxyConfig = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task
    Function UploadRemoteFile(Url As String, DestinationFolder As String, Optional FileName As String = Nothing, Optional DisableRedirects As Boolean = False, Optional Fields As Fields = Nothing) As Task(Of JSONLink)
    Function UnPublishFile(FilePath As String) As Task(Of JSONLink)
    Function PublishFile(FilePath As String) As Task(Of JSONLink)
    Function RenameFile(SourceFile As String, RenameTo As String) As Task(Of JSONLink)
    Function CheckIfFileExists(FilePath As String) As Task(Of Boolean)
    Function PreviewFile(PreviewURL As Uri, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional _proxi As ProxyConfig = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task(Of IO.Stream)


#End Region

#Region "folder"
    Function PublishFolder(FolderPath As String) As Task(Of JSONLink)
    Function DeleteFolder(FolderPath As String, Optional Fields As Fields = Nothing) As Task(Of JSONLink)
    Function TrashFolder(FolderPath As String, Optional Fields As Fields = Nothing) As Task(Of JSONLink)
    Function MoveFolder(SourceFolder As String, DestinationFolder As String, FolderName As String, Optional Fields As Fields = Nothing, Optional Overwriting As Boolean = False) As Task(Of JSONLink)
    Function CopyFolder(SourceFolder As String, DestinationFolder As String, FolderName As String, Optional Fields As Fields = Nothing, Optional Overwriting As Boolean = False) As Task(Of JSONLink)
    Function DownloadFolderAsZip(FullFilePath As String, FileSaveDir As String, FileName As String, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional _proxi As ProxyConfig = Nothing, Optional TimeOut As Integer = 60, Optional token As Threading.CancellationToken = Nothing) As Task
    Function UnPublishFolder(FolderPath As String) As Task(Of JSONLink)
    Function RenameFolder(SourceFolder As String, RenameTo As String) As Task(Of JSONLink)
    Function CreateDir(DestinationFolder As String, FolderName As String, Optional Fields As Fields = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task(Of JSONLink)
    Function CheckIfFolderExists(FolderPath As String) As Task(Of Boolean)

#End Region

#Region "Files And Folders"
    Function ListAllFilesAndFolders(Path As String, Optional Fields As Fields? = Nothing, Optional Limit As String = Nothing, Optional Offset As String = Nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As String = Nothing, Optional Sort As Sort? = Nothing) As Task(Of JSONFilesandFolders)
    Function TrashFileFolder(FileFolderPath As String, Optional Fields As Fields = Nothing) As Task(Of JSONLink)
    Function MoveAndRenameFileFolder(SourceFileFolder As String, DestinationFolder As String, FileName As String, Optional Overwriting As Boolean = False) As Task(Of JSONLink)
    Function CopyAndRenameFileFolder(SourceFileFolder As String, DestinationFolder As String, FolderName As String, Optional Overwriting As Boolean = False) As Task(Of JSONLink)
    Function PublishFileFolder(SourceFileFolder As String) As Task(Of JSONLink)
    Function UnPublishFileFolder(SourceFileFolder As String) As Task(Of JSONLink)
    Function GetFileFolderMetadata(SourceFileFolder As String, Optional PreviewCrop As Boolean = True, Optional PreviewSize As String = Nothing) As Task(Of JSONFilesandFolders)
    Function CheckIfFileFolderExists(SourceFileFolder As String) As Task(Of Boolean)
    Function RenameFileFolder(SourceFileFolder As String, RenameTo As String) As Task(Of JSONLink)

#End Region

    Function ListAllSharedLinksOfUser(Optional Fields As Fields = Nothing, Optional Limit As String = Nothing, Optional Offset As String = Nothing, Optional FilteredBy As JSONFilesandFolders.File_or_Folder? = Nothing, Optional PreviewSize As String = Nothing) As Task(Of JSON_Embedded)
    Function UserInfo(Optional token As Threading.CancellationToken = Nothing) As Task(Of JSONuserinfo)
    Function AddMetaForFile(Path As String, Optional Fields As Fields = Nothing) As Task(Of JSONFilesandFolders)
    Function ListLatestUploadedFiles(Optional MediaType As Media_Type? = Nothing, Optional Limit As Integer = 20, Optional Fields As Fields = Nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As String = Nothing) As Task(Of JSON_Embedded)
    
    Function GetDownloadUrl(SourceFile As String) As Task(Of JSONLink)
    
    Function EmptyTrash(Optional disk_File_Dir_and_Name As String = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task(Of JSONyandex.JSONLink)
    Function RestoringFileorFolderFromTrash(disk_File_Dir_and_Name As String, Optional TheNewNameofTheResource As String = Nothing, Optional Overwriting As Boolean = False, Optional token As Threading.CancellationToken = Nothing) As Task(Of JSONyandex.JSONLink)

    Function CopyingMovingDeletingUrluploadingOperationStatus(OperationHref As String) As Task(Of JSON_CopyingMovingDeletingUrluploadingOperationStatus)

    'Function TrashMultipleFileFolder(FileFolderPaths As List(Of String)) As Task(Of Boolean)
    Function ListPublicLink(PublicFolderUrlOrKey As String, Optional FilePathInsidePublicFolder As String = Nothing, Optional Sort As Sort? = Nothing, Optional Limit As String = Nothing, Optional Offset As String = Nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As String = Nothing) As Task(Of JSONFilesandFolders)

    Function GetDownloadUrlOfFileInPublicFolder(PublicFolderUrlOrKey As String, Optional FilePathInsidePublicFolder As String = Nothing) As Task(Of JSONLink)
    Function DownloadFileFromPublicFolder(FolderPublicLink As String, FilePathToDownload As String, OutputDirPath As String, Optional DownloadUrl As String = Nothing, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional _proxi As ProxyConfig = Nothing, Optional TimeOut As Integer = 60, Optional token As Threading.CancellationToken = Nothing) As Task(Of Boolean)
    

End Interface

File: IClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using YandexSDK.JSONyandex;

namespace YandexSDK
{
	public interface IClient
	{
		Task<JSONAllFiles> ListAllFiles(List<string> MediaType = null, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, int Limit = 20, string Offset = null, bool PreviewCrop = true, string PreviewSize = null, YdXutilities.Sort Sort = YdXutilities.Sort.nothing);

		Task<JSONLink> CopyFile(string SourceFile, string DestinationFolder, string FileName, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, bool Overwriting = false);

		Task<JSONLink> MoveFile(string SourceFile, string DestinationFolder, string FileName, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, bool Overwriting = false);

		Task<JSONLink> DeleteFile(string FilePath, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task<JSONLink> TrashFile(string FilePath, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task DownloadFile(string FullFilePath, string FileSaveDir, string FileName, IProgress<ReportStatus> ReportCls = null, ProxyConfig _proxi = null, int TimeOut = 60, CancellationToken token = default(CancellationToken));

		Task<Stream> DownloadFileAsStream(string FullFilePath, IProgress<ReportStatus> ReportCls = null, ProxyConfig _proxi = null, int TimeOut = 60, CancellationToken token = default(CancellationToken));

		Task UploadLocalFile(object FileToUpload, string FileName, string DestinationFolder, YClient.UploadTypes UploadType, bool Overwriting = false, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, IProgress<ReportStatus> ReportCls = null, ProxyConfig _proxi = null, CancellationToken token = default(CancellationToken));

		Task<JSONLink> UploadRemoteFile(string Url, string DestinationFolder, string FileName = null, bool DisableRedirects = false, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task<JSONLink> UnPublishFile(string FilePath);

		Task<JSONLink> PublishFile(string FilePath);

		Task<JSONLink> RenameFile(string SourceFile, string RenameTo);

		Task<bool> CheckIfFileExists(string FilePath);

		Task<Stream> PreviewFile(Uri PreviewURL, IProgress<ReportStatus> ReportCls = null, ProxyConfig _proxi = null, CancellationToken token = default(CancellationToken));

		Task<JSONLink> PublishFolder(string FolderPath);

		Task<JSONLink> DeleteFolder(string FolderPath, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task<JSONLink> TrashFolder(string FolderPath, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task<JSONLink> MoveFolder(string SourceFolder, string DestinationFolder, string FolderName, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, bool Overwriting = false);

		Task<JSONLink> CopyFolder(string SourceFolder, string DestinationFolder, string FolderName, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, bool Overwriting = false);

		Task DownloadFolderAsZip(string FullFilePath, string FileSaveDir, string FileName, IProgress<ReportStatus> ReportCls = null, ProxyConfig _proxi = null, int TimeOut = 60, CancellationToken token = default(CancellationToken));

		Task<JSONLink> UnPublishFolder(string FolderPath);

		Task<JSONLink> RenameFolder(string SourceFolder, string RenameTo);

		Task<JSONLink> CreateDir(string DestinationFolder, string FolderName, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, CancellationToken token = default(CancellationToken));

		Task<bool> CheckIfFolderExists(string FolderPath);

		Task<JSONFilesandFolders> ListAllFilesAndFolders(string Path, YdXutilities.Fields? Fields = null, int Limit = 20, int Offset = 0, bool PreviewCrop = true, string PreviewSize = null, YdXutilities.Sort? Sort = null);

		Task<JSONLink> TrashFileFolder(string FileFolderPath, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task<JSONLink> MoveAndRenameFileFolder(string SourceFileFolder, string DestinationFolder, string FileName, bool Overwriting = false);

		Task<JSONLink> CopyAndRenameFileFolder(string SourceFileFolder, string DestinationFolder, string FolderName, bool Overwriting = false);

		Task<JSONLink> PublishFileFolder(string SourceFileFolder);

		Task<JSONLink> UnPublishFileFolder(string SourceFileFolder);

		Task<JSONFilesandFolders> GetFileFolderMetadata(string SourceFileFolder, bool PreviewCrop = true, string PreviewSize = null);

		Task<bool> CheckIfFileFolderExists(string SourceFileFolder);

		Task<JSONLink> RenameFileFolder(string SourceFileFolder, string RenameTo);

		Task<JSON_Embedded> ListAllSharedLinksOfUser(YdXutilities.Fields Fields = YdXutilities.Fields.nothing, string Limit = null, string Offset = null, JSONFilesandFolders.File_or_Folder? FilteredBy = null, string PreviewSize = null);

		Task<JSONuserinfo> UserInfo(CancellationToken token = default(CancellationToken));

		Task<JSONFilesandFolders> AddMetaForFile(string Path, YdXutilities.Fields Fields = YdXutilities.Fields.nothing);

		Task<JSON_Embedded> ListLatestUploadedFiles(YdXutilities.Media_Type? MediaType = null, int Limit = 20, YdXutilities.Fields Fields = YdXutilities.Fields.nothing, bool PreviewCrop = true, string PreviewSize = null);

		Task<JSONLink> GetDownloadUrl(string SourceFile);

		Task<JSONLink> EmptyTrash(string disk_File_Dir_and_Name = null, CancellationToken token = default(CancellationToken));

		Task<JSONLink> RestoringFileorFolderFromTrash(string disk_File_Dir_and_Name, string TheNewNameofTheResource = null, bool Overwriting = false, CancellationToken token = default(CancellationToken));

		Task<JSON_CopyingMovingDeletingUrluploadingOperationStatus> CopyingMovingDeletingUrluploadingOperationStatus(string OperationHref);

		Task<JSONFilesandFolders> ListPublicLink(string PublicFolderUrlOrKey, string FilePathInsidePublicFolder = null, YdXutilities.Sort? Sort = null, string Limit = null, string Offset = null, bool PreviewCrop = true, string PreviewSize = null);

		Task<JSONLink> GetDownloadUrlOfFileInPublicFolder(string PublicFolderUrlOrKey, string FilePathInsidePublicFolder = null);

		Task<bool> DownloadFileFromPublicFolder(string FolderPublicLink, string FilePathToDownload, string OutputDirPath, string DownloadUrl = null, IProgress<ReportStatus> ReportCls = null, ProxyConfig _proxi = null, int TimeOut = 60, CancellationToken token = default(CancellationToken));
	}
}
